module1-introduction-to-backend/r2.1-diving-into-rest-apis/README.md

```diff
@@ -0,0 +1,122 @@
+1. Understanding HTTP request types
+2. Understanding HTTP request and response structure
+3. Understanding HTTP response status codes
+
+## HTTP Requests
+In the previous lesson, we briefly talked about GET, POST, PUT and DELETE request methods.
+
+HTTP defines a set of request methods to indicate the desired action to be performed for a given resource. This set includes:
+- GET: Requests a representation of the specified resource. Requests using GET should only retrieve data.
+- HEAD: Asks for a response identical to that of a GET request, but without the response body.
+- POST: Used to submit an entity to the specified resource, often causing a change in state or side effects on the server.
+- PUT: Replaces all current representations of the target resource with the request payload.
+- DELETE: Requests to delete the specified resource.
+- CONNECT: Establishes a tunnel to the server identified by the target resource.
+- OPTIONS: Used to describe the communication options for the target resource.
+- TRACE: Performs a message loop-back test along the path to the target resource.
+- PATCH: Used to apply partial modifications to a resource.
+
+While developing APIs, GET and POST are the most commonly and frequently used methods.
+
+## Structure of a HTTP Request
+
+HTTP requests, and responses, share similar structure and are composed of:
+1. A start-line describing the requests to be implemented, or its status of whether successful or a failure. This start-line is always a single line.
+2. An optional set of HTTP headers specifying the request, or describing the body included in the message.
+3. A blank line indicating all meta-information for the request has been sent.
+4. An optional body containing data associated with the request (like content of an HTML form), or the document associated with a response. The presence of the body and its size is specified by the start-line and HTTP headers.
+
+### Start line
+HTTP requests are messages sent by the client to initiate an action on the server. Their start-line contain three elements:
+1. An **HTTP method**, a verb (like GET, PUT or POST) or a noun (like HEAD or OPTIONS), that describes the action to be performed. For example, GET indicates that a resource should be fetched or POST means that data is pushed to the server.
+2. The **HTTP version**, which defines the structure of the remaining message, acting as an indicator of the expected version to use for the response.
+3. The **request target**, usually a URL, or the absolute path of the protocol, port, and domain are usually characterized by the request context. The format of this request target varies between different HTTP methods. It can be
+- An absolute path, ultimately followed by a '?' and query string. This is the most common form, known as the origin form, and is used with GET, POST, HEAD, and OPTIONS methods.
+```
+POST / HTTP/1.1 GET /background.png HTTP/1.0 HEAD /test.html?query=alibaba HTTP/1.1 OPTIONS /anypage.html HTTP/1.0
```

Comment:
i would move this example or include an example either at line 24 or line 29

Reply:
All these examples actually come under point 3, I missed adding the indentation.